assets/app/scripts/services/data.js

```diff
@@ -514,55 +521,55 @@ DataService.prototype.createStream = function(resource, name, context, opts, isR
 //                    a websocket. Default is false.
 //            pollInterval: in milliseconds, how long to wait between polling the server
 //                    only applies if poll=true.  Default is 5000.
-//
+//            http:   similar to .get, etc. at this point, only used to pass http.params for filtering
 // returns handle to the watch, needed to unwatch e.g.
 //        var handle = DataService.watch(resource,context,callback[,opts])
 //        DataService.unwatch(handle)
   DataService.prototype.watch = function(resource, context, callback, opts) {
     resource = APIService.toResourceGroupVersion(resource);
     opts = opts || {};
-
+    var key = this._uniqueKeyForResourceContext(resource, null, context, _.get(opts, 'http.params'));
     if (callback) {
       // If we were given a callback, add it
-      this._watchCallbacks(resource, context).add(callback);
+      this._watchCallbacks(key).add(callback);
     }
-    else if (!this._watchCallbacks(resource, context).has()) {
+    else if (!this._watchCallbacks(key).has()) {
       // We can be called with no callback in order to re-run a list/watch sequence for existing callbacks
       // If there are no existing callbacks, return
       return {};
     }
 
-    var existingWatchOpts = this._watchOptions(resource, context);
+    var existingWatchOpts = this._watchOptions(key);
     if (existingWatchOpts) {
       // Check any options for compatibility with existing watch
       if (existingWatchOpts.poll != opts.poll) {
         throw "A watch already exists for " + resource + " with a different polling option.";
       }
     }
     else {
-      this._watchOptions(resource, context, opts);
+      this._watchOptions(key, opts);
     }
 
     var self = this;
 
-    if (this._watchInFlight(resource, context) && this._resourceVersion(resource, context)) {
+    if (this._watchInFlight(key) && this._resourceVersion(key)) {
       if (callback) {
         $timeout(function() {
-          callback(self._data(resource, context));
+          callback(self._data(key));
```

Comment:
the logic here assumed that if a watch was in flight and the initial resource version existed then we had complete / accurate data, but if there is a cache miss here (introduced by the LRU) then we can no longer assume we have the right data, we need to reinitiate the whole watch operation in that case (including re-listing first)